en/02_Developer_Guides/03_Forms/01_Validation.md

```diff
@@ -71,71 +71,94 @@ In this example we will be required to input a value for `Name` and a valid emai
     {
         if (str_ends_with($this->owner->Value(), '@example.com')) {
-            $validator->validationError($this->owner->Name(), 'Please provide a valid email address');
-            $result = false;
+            $result->addFieldError(
+                $this->owner->Name(),
+                'Please provide a valid email address which does not end with @example.com'
+            );
         }
     }
 }
 ```
 
-> [!WARNING]
-> This extension hook will not work without the ampersand (`&`) in the `&$result` argument. This is because the return
-> value of the function is ignored, so the validation result has to be updated by changing the value of the `$result`
-> variable. This is known as [passing by reference](https://www.php.net/manual/en/language.references.pass.php).
-
 ## Validation in `FormField` subclasses
 
-Subclasses of `FormField` can define their own version of `validate` to provide custom validation rules such as the
-above example with the `Email` validation. The `validate` method on `FormField` takes a single argument of the current
-`Validator` instance.
+Subclasses of `FormField` can define their own version of `validate()` to provide custom validation rules such as the
+above example with the `Email` validation.
 
 ```php
 namespace App\Form\Field;
 
+use SilverStripe\Core\Validation\ValidationResult;
 use SilverStripe\Forms\NumericField;
 
 class CustomNumberField extends NumericField
 {
     // ...
 
-    public function validate($validator)
+    public function validate(): ValidationResult
     {
-        if ((int) $this->Value() === 10) {
-            $validator->validationError($this->Name(), 'This value cannot be 10');
-            return $this->extendValidationResult(false, $validator);
-        }
-
-        return $this->extendValidationResult(true, $validator);
+        $result = ValidationResult::create();
+        $this->beforeExtending('updateValidate', function () use ($result) {
+            if ((int) $this->Value() === 10) {
+                $result->addFieldError($this->Name(), 'This value cannot be 10');
+            }
+        });
+        return $result->combineAnd(parent::validate());
     }
-}
 ```
 
-The `validate` method should compute a boolean (`true` if the value passes validation and `false` if Silverstripe CMS
-should trigger a validation error on the page) and pass this to the `extendValidationResult` method to allow extensions
-to hook into the validation logic. In addition, in the event of failed validation, a useful error message must be set
-on the given validator.
+The `validate()` method must create a [`ValidationResult`](api:SilverStripe\Core\Validation\ValidationResult) object and add any errors to it, which should contain a useful error message. The `validate()` method should utilise the `$this->beforeExtending('updateValidate', ...` block and be followed by `return $result->combineAnd(parent::validate());` in order for code in the parent class to be executed in the correct order.
```

Comment:
It doesn't have to create one, but it must return one.
Best practice if we're using `beforeExtending()` is to use the one passed into the extension hook, which the parent class creates.

Reply:
Updated